# <font color='blue'>Pandas:</font>

* <font color='b'>Pandas comes from the econometrics term 'Panel Data'</font>

* <font color='purple'>Functions for analyzing, cleaning, exploring, and manipulating data.</font>
* <font color='brown'>Pandas stands for "Python Data Analysis Library."</font>

 <font color='magenta'> ================================= </font>



#Recall:

Syntax for renaming columns:

<font color ='red'>`df.rename(columns={'old_name': 'new_name', ...}, inplace=True)`

Syntax: Add a New Column Using pd.cut()

<font color ='red'>`df['new_column'] = pd.cut(df['column_to_bin'], bins, labels=labels, right =True)`

Syntax for converting column entries to uppercase

<font color ='red'>`df['column_name'] = df['column_name'].str.upper()`

<h2>Day 4: Advanced Data Manipulation and Visualization</h2>

<h3>1. Filtering and Querying:</h3>
<ul>
    <li><strong>Boolean Indexing for Condition-Based Filtering:</strong>
        <ul>
            <li>For a single column: <code><font color="blue">df[df['column'] > value]</font></code></li>
            <li>For multiple columns: <code><font color="blue">df[(df['column1'] > value1) &amp; (df['column2'] < value2)]</font></code></li>
            <li>Using the <code>.isin()</code> method for filtering rows based on a list of values: <code><font color="blue">df[df['column'].isin([value1, value2, ...])]</font></code></li>
        </ul>
    </li>
    <li><strong>Using <code>.loc</code> and <code>.iloc</code> for Precise Data Selection:</strong>
        <ul>
            <li><code>.loc</code> syntax: <code><font color="blue">df.loc[row_label, column_label]</font></code></li>
            <li><code>.iloc</code> syntax: <code><font color="blue">df.iloc[row_index, column_index]</font></code></li>
        </ul>
    </li>
    <li><strong>Flexible Data Queries with <code>.query()</code>:</strong>
        <ul>
            <li>For a single condition: <code><font color="blue">df.query('expression')</font></code></li>
            <li>For multiple conditions: <code><font color="blue">df.query('expression1 and/or expression2')</font></code></li>
        </ul>
    </li>
</ul>

<h3>2. Sorting and Grouping:</h3>
<ul>
    <li><strong>Ordering Datasets with <code>.sort_values()</code>:</strong>
        <ul>
            <li>By a single column: <code><font color="blue">df.sort_values(by='column', ascending=True)</font></code></li>
            <li>By multiple columns: <code><font color="blue">df.sort_values(by=['column1', 'column2'], ascending=[True, False])</font></code></li>
        </ul>
    </li>
    <li><strong>Grouping Data with <code>.groupby()</code> and Exploring Basic Aggregations:</strong>
        <ul>
            <li>Simple grouping: <code><font color="blue">df.groupby('column')</font></code></li>
            <li>Multi-column grouping: <code><font color="blue">df.groupby(['column1', 'column2'])</font></code></li>
            <li>Aggregation examples:
                <ul>
                    <li>Mean: <code><font color="blue">df.groupby('column').mean()</font></code></li>
                    <li>Custom aggregations using <code>.agg()</code>: <code><font color="blue">df.groupby(['column1', 'column2']).agg({'column3': 'mean', 'column4': 'sum'})</font></code></li>
                </ul>
            </li>
        </ul>
    </li>
</ul>

<h3>3. Handling Missing Data:</h3>
<ul>
    <li><strong>Utilizing <code>.fillna()</code> to Fill Missing Entries:</strong>
        <ul>
            <li>For the entire DataFrame: <code><font color="blue">df.fillna(value)</font></code></li>
            <li>For specific columns: <code><font color="blue">df.fillna({'column1': value1, 'column2': value2})</font></code></li>
        </ul>
    </li>
    <li><strong>Employing <code>.dropna()</code> to Clean the Dataset:</strong>
        <ul>
            <li>Dropping rows with any missing data: <code><font color="blue">df.dropna(axis=0, how='any')</font></code></li>
            <li>Dropping columns where all data is missing: <code><font color="blue">df.dropna(axis=1, how='all')</font></code></li>
        </ul>
    </li>
</ul>


In [1]:
# from google.colab import drive
# drive.mount("/content/my_drive")

In [2]:
# Imports
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import os

In [3]:
# load the .csv called "Pakistan_company_data.csv"
data = pd.read_csv("Pakistan_company_data.csv")
data.head()

Employee_ID  Age  Salary Department   Join_Date Last_Performance_Rating  \
0      PK-001   58   56736         IT  2005-01-02               Excellent   
1      PK-002   48  142859  Marketing  2005-01-09                    Good   
2      PK-003   34  142181         IT  2005-01-16                    Poor   
3      PK-004   62  161926  Logistics  2005-01-23               Excellent   
4      PK-005   27  120084    Finance  2005-01-30                    Good   

   Days_Absent  
0          5.0  
1          2.0  
2          NaN  
3          NaN  
4         24.0

In [4]:
data.describe()

Age         Salary  Days_Absent
count  200.000000     200.000000   160.000000
mean    42.885000  116007.670000    12.481250
std     13.491176   48077.736033     8.695384
min     20.000000   32368.000000     0.000000
25%     31.000000   73514.750000     5.000000
50%     43.500000  123616.000000    11.000000
75%     55.000000  155060.750000    19.250000
max     64.000000  199695.000000    29.000000

### 1. Filtering and Querying:



Write a query to find all employees in the 'Finance' department.

In [5]:
df = data[data["Department"] == 'Finance']
df

Employee_ID  Age  Salary Department   Join_Date Last_Performance_Rating  \
4        PK-005   27  120084    Finance  2005-01-30                    Good   
6        PK-007   58   37392    Finance  2005-02-13                 Average   
32       PK-033   35  163883    Finance  2005-08-14                    Good   
35       PK-036   22  154249    Finance  2005-09-04                     NaN   
67       PK-068   53   37400    Finance  2006-04-16                    Good   
76       PK-077   59   42183    Finance  2006-06-18             Outstanding   
80       PK-081   37   62711    Finance  2006-07-16                    Good   
85       PK-086   60  163272    Finance  2006-08-20                 Average   
86       PK-087   48   51357    Finance  2006-08-27                    Good   
89       PK-090   20   32869    Finance  2006-09-17             Outstanding   
93       PK-094   43  120272    Finance  2006-10-15             Outstanding   
101      PK-102   54  116416    Finance  2006-12-10                    Good   
103      PK-104   52   74064    Finance  2006-12-24               Excellent   
112      PK-113   31  166967    Finance  2007-02-25                    Good   
119      PK-120   63  166672    Finance  2007-04-15                 Average   
130      PK-131   24   73525    Finance  2007-07-01                 Average   
132      PK-133   33   93208    Finance  2007-07-15             Outstanding   
142      PK-143   58  145668    Finance  2007-09-23                    Poor   
152      PK-153   48   34014    Finance  2007-12-02                    Good   
153      PK-154   55   41093    Finance  2007-12-09                 Average   
159      PK-160   21  188823    Finance  2008-01-20                    Poor   
165      PK-166   63  180262    Finance  2008-03-02                    Good   
169      PK-170   30  140627    Finance  2008-03-30                    Good   
173      PK-174   52   36655    Finance  2008-04-27                    Good   
185      PK-186   52   52677    Finance  2008-07-20                     NaN   
187      PK-188   63   61024    Finance  2008-08-03                     NaN   
188      PK-189   43  100313    Finance  2008-08-10               Excellent   

     Days_Absent  
4           24.0  
6            4.0  
32           1.0  
35           7.0  
67          23.0  
76           NaN  
80          27.0  
85          21.0  
86          16.0  
89           5.0  
93           1.0  
101          4.0  
103         25.0  
112         14.0  
119         10.0  
130          NaN  
132         11.0  
142          8.0  
152         29.0  
153         27.0  
159         10.0  
165          0.0  
169          NaN  
173          4.0  
185         29.0  
187         24.0  
188          NaN

Write a query to find all employees younger than 30 years who also earn more than 100,000 PKR.


In [6]:
df2 = data[(data["Age"] < 30) & (data["Salary"] > 100000)]
df2

Employee_ID  Age  Salary Department   Join_Date Last_Performance_Rating  \
4        PK-005   27  120084    Finance  2005-01-30                    Good   
15       PK-016   22  173760  Marketing  2005-04-17                    Good   
17       PK-018   21  198229         HR  2005-05-01               Excellent   
22       PK-023   21  140078  Logistics  2005-06-05                 Average   
35       PK-036   22  154249    Finance  2005-09-04                     NaN   
37       PK-038   26  172483  Logistics  2005-09-18                    Good   
39       PK-040   28  169752      Admin  2005-10-02                     NaN   
42       PK-043   23  130689      Admin  2005-10-23                 Average   
45       PK-046   28  162874  Logistics  2005-11-13               Excellent   
58       PK-059   23  177443         IT  2006-02-12                 Average   
59       PK-060   21  141472         IT  2006-02-19                    Good   
60       PK-061   25  163121  Logistics  2006-02-26               Excellent   
62       PK-063   23  133727  Marketing  2006-03-12                    Good   
68       PK-069   29  147858         IT  2006-04-23                 Average   
73       PK-074   27  150885         IT  2006-05-28                    Good   
91       PK-092   26  142296         HR  2006-10-01               Excellent   
92       PK-093   28  124179         IT  2006-10-08             Outstanding   
96       PK-097   27  121412      Admin  2006-11-05                 Average   
100      PK-101   27  158391      Admin  2006-12-03                 Average   
104      PK-105   24  137450         HR  2006-12-31               Excellent   
110      PK-111   28  164508  Logistics  2007-02-11               Excellent   
111      PK-112   27  106213         IT  2007-02-18             Outstanding   
124      PK-125   20  135983      Admin  2007-05-20                 Average   
128      PK-129   22  112989      Admin  2007-06-17                     NaN   
144      PK-145   23  148015  Logistics  2007-10-07             Outstanding   
159      PK-160   21  188823    Finance  2008-01-20                    Poor   
162      PK-163   25  135510  Marketing  2008-02-10                 Average   
174      PK-175   20  100031      Admin  2008-05-04                    Good   
180      PK-181   25  154862         IT  2008-06-15                    Good   
181      PK-182   27  148834         IT  2008-06-22             Outstanding   
196      PK-197   21  117235      Admin  2008-10-05                    Poor   
197      PK-198   22  193165  Logistics  2008-10-12                 Average   

     Days_Absent  
4           24.0  
15           NaN  
17           NaN  
22           3.0  
35           7.0  
37          11.0  
39           9.0  
42           5.0  
45           1.0  
58           5.0  
59          22.0  
60          15.0  
62          29.0  
68          22.0  
73           7.0  
91          29.0  
92           NaN  
96           7.0  
100         11.0  
104          NaN  
110         16.0  
111         11.0  
124          NaN  
128          9.0  
144         14.0  
159         10.0  
162         11.0  
174         13.0  
180          4.0  
181         18.0  
196         13.0  
197         18.0

The company operates multiple departments such as 'IT', 'Marketing', and 'Finance'. Use the `.isin()` method to filter and print the data for employees who belong to these specific departments.

In [7]:
dept_chosen= ['IT', 'Marketing', 'Finance']
data_chosen = data[data['Department'].isin(dept_chosen)]
data_chosen



Employee_ID  Age  Salary Department   Join_Date Last_Performance_Rating  \
0        PK-001   58   56736         IT  2005-01-02               Excellent   
1        PK-002   48  142859  Marketing  2005-01-09                    Good   
2        PK-003   34  142181         IT  2005-01-16                    Poor   
4        PK-005   27  120084    Finance  2005-01-30                    Good   
6        PK-007   58   37392    Finance  2005-02-13                 Average   
..          ...  ...     ...        ...         ...                     ...   
189      PK-190   34  157813         IT  2008-08-17                 Average   
190      PK-191   51  125285  Marketing  2008-08-24                    Good   
192      PK-193   43  143632         IT  2008-09-07             Outstanding   
193      PK-194   60  180580         IT  2008-09-14               Excellent   
194      PK-195   31   33051  Marketing  2008-09-21               Excellent   

     Days_Absent  
0            5.0  
1            2.0  
2            NaN  
4           24.0  
6            4.0  
..           ...  
189          9.0  
190         24.0  
192          7.0  
193          7.0  
194          NaN  

[99 rows x 7 columns]

Using `.loc`, retrieve the details of the employee with the ID 'PK-050'.

In [8]:
df3 = data.loc[data['Employee_ID'] == 'PK-050']
df3

Employee_ID  Age  Salary Department   Join_Date Last_Performance_Rating  \
49      PK-050   47   77254  Marketing  2005-12-11                    Good   

    Days_Absent  
49         12.0

Using `.iloc`, retrieve the details of the 10th to 15th employees in the dataframe.

In [9]:
df4= data.iloc[10:16]
df4

Employee_ID  Age  Salary Department   Join_Date Last_Performance_Rating  \
10      PK-011   30  187381      Admin  2005-03-13                     NaN   
11      PK-012   43  100467         IT  2005-03-20                 Average   
12      PK-013   55  128506      Admin  2005-03-27                    Good   
13      PK-014   59  139751  Logistics  2005-04-03                 Average   
14      PK-015   43  148906         HR  2005-04-10                    Good   
15      PK-016   22  173760  Marketing  2005-04-17                    Good   

    Days_Absent  
10         24.0  
11          NaN  
12          6.0  
13          NaN  
14         25.0  
15          NaN

Use `.query()` to select employees with an 'Excellent' performance rating.

In [10]:
df5 = data.query('Last_Performance_Rating == "Excellent"')
df5

Employee_ID  Age  Salary Department   Join_Date Last_Performance_Rating  \
0        PK-001   58   56736         IT  2005-01-02               Excellent   
3        PK-004   62  161926  Logistics  2005-01-23               Excellent   
5        PK-006   40  199695         HR  2005-02-06               Excellent   
9        PK-010   30  155657         IT  2005-03-06               Excellent   
17       PK-018   21  198229         HR  2005-05-01               Excellent   
20       PK-021   49  119045         HR  2005-05-22               Excellent   
28       PK-029   44   81663      Admin  2005-07-17               Excellent   
41       PK-042   37  101295  Marketing  2005-10-16               Excellent   
44       PK-045   33   67504  Logistics  2005-11-06               Excellent   
45       PK-046   28  162874  Logistics  2005-11-13               Excellent   
46       PK-047   45   38155  Logistics  2005-11-20               Excellent   
55       PK-056   36  150975  Logistics  2006-01-22               Excellent   
57       PK-058   59  137512      Admin  2006-02-05               Excellent   
60       PK-061   25  163121  Logistics  2006-02-26               Excellent   
61       PK-062   61  192688      Admin  2006-03-05               Excellent   
63       PK-064   48  142893         IT  2006-03-19               Excellent   
66       PK-067   63   59855         IT  2006-04-09               Excellent   
69       PK-070   55   45151         HR  2006-04-30               Excellent   
70       PK-071   33   96690      Admin  2006-05-07               Excellent   
79       PK-080   64  146336         IT  2006-07-09               Excellent   
91       PK-092   26  142296         HR  2006-10-01               Excellent   
102      PK-103   54  100271         IT  2006-12-17               Excellent   
103      PK-104   52   74064    Finance  2006-12-24               Excellent   
104      PK-105   24  137450         HR  2006-12-31               Excellent   
105      PK-106   61   49830         IT  2007-01-07               Excellent   
106      PK-107   58  178501      Admin  2007-01-14               Excellent   
110      PK-111   28  164508  Logistics  2007-02-11               Excellent   
137      PK-138   34  161373      Admin  2007-08-19               Excellent   
148      PK-149   58  184165  Marketing  2007-11-04               Excellent   
156      PK-157   26   86958  Marketing  2007-12-30               Excellent   
163      PK-164   47   84748  Logistics  2008-02-17               Excellent   
172      PK-173   58  107575  Marketing  2008-04-20               Excellent   
179      PK-180   58  172038      Admin  2008-06-08               Excellent   
188      PK-189   43  100313    Finance  2008-08-10               Excellent   
193      PK-194   60  180580         IT  2008-09-14               Excellent   
194      PK-195   31   33051  Marketing  2008-09-21               Excellent   

     Days_Absent  
0            5.0  
3            NaN  
5            NaN  
9            6.0  
17           NaN  
20          19.0  
28          14.0  
41           4.0  
44           NaN  
45           1.0  
46          26.0  
55           NaN  
57           5.0  
60          15.0  
61           NaN  
63           NaN  
66          29.0  
69           NaN  
70           NaN  
79           NaN  
91          29.0  
102          5.0  
103         25.0  
104          NaN  
105          3.0  
106         18.0  
110         16.0  
137         19.0  
148          2.0  
156         13.0  
163          2.0  
172          8.0  
179          4.0  
188          NaN  
193          7.0  
194          NaN

Use `.query()` to find employees in either 'HR' or 'Marketing' departments who have 'Poor' or 'Average' performance ratings.

In [20]:
df6 = data.query('Department == ["HR", "Marketing"] and Last_Performance_Rating == ["Poor", "Average"]')
df6 = data.query('Department in ["HR", "Marketing"] and Last_Performance_Rating in ["Poor", "Average"]')  #This is same thing just using 'in' instead of == 
df6

Employee_ID  Age  Salary Department   Join_Date Last_Performance_Rating  \
7        PK-008   38   85680  Marketing  2005-02-20                    Poor   
24       PK-025   52   52671         HR  2005-06-19                 Average   
27       PK-028   63  116202         HR  2005-07-10                 Average   
74       PK-075   33  156071         HR  2006-06-04                 Average   
77       PK-078   40  190371  Marketing  2006-06-25                    Poor   
78       PK-079   35  173946  Marketing  2006-07-02                 Average   
81       PK-082   43   35539  Marketing  2006-07-23                 Average   
88       PK-089   64  107505  Marketing  2006-09-10                 Average   
94       PK-095   20   68467  Marketing  2006-10-22                 Average   
114      PK-115   52   60746  Marketing  2007-03-11                    Poor   
125      PK-126   54  154123         HR  2007-05-27                    Poor   
129      PK-130   20   66212         HR  2007-06-24                 Average   
131      PK-132   45  193707         HR  2007-07-08                 Average   
133      PK-134   58   63828  Marketing  2007-07-22                 Average   
139      PK-140   61  154019         HR  2007-09-02                 Average   
143      PK-144   51   36102         HR  2007-09-30                 Average   
146      PK-147   56  195656         HR  2007-10-21                 Average   
149      PK-150   64  133746         HR  2007-11-11                 Average   
155      PK-156   51   65777  Marketing  2007-12-23                 Average   
162      PK-163   25  135510  Marketing  2008-02-10                 Average   
164      PK-165   47   35801  Marketing  2008-02-24                    Poor   
171      PK-172   44   32368         HR  2008-04-13                    Poor   
177      PK-178   60  174475         HR  2008-05-25                 Average   
191      PK-192   51  176410         HR  2008-08-31                 Average   
195      PK-196   58  144692         HR  2008-09-28                 Average   

     Days_Absent  
7            NaN  
24           2.0  
27          17.0  
74           4.0  
77           5.0  
78           NaN  
81          10.0  
88           NaN  
94          14.0  
114         25.0  
125          NaN  
129          4.0  
131          NaN  
133          1.0  
139          4.0  
143          NaN  
146          5.0  
149          NaN  
155         17.0  
162         11.0  
164          2.0  
171         10.0  
177          2.0  
191         17.0  
195          7.0

### 2. Sorting and Grouping:

Sort the dataset by 'Salary' in ascending order.

In [12]:
df7 = data.sort_values(by='Salary', ascending=True)
df7

Employee_ID  Age  Salary Department   Join_Date Last_Performance_Rating  \
171      PK-172   44   32368         HR  2008-04-13                    Poor   
89       PK-090   20   32869    Finance  2006-09-17             Outstanding   
194      PK-195   31   33051  Marketing  2008-09-21               Excellent   
135      PK-136   28   33420         HR  2007-08-05                    Good   
152      PK-153   48   34014    Finance  2007-12-02                    Good   
..          ...  ...     ...        ...         ...                     ...   
131      PK-132   45  193707         HR  2007-07-08                 Average   
147      PK-148   42  193817  Logistics  2007-10-28                    Good   
146      PK-147   56  195656         HR  2007-10-21                 Average   
17       PK-018   21  198229         HR  2005-05-01               Excellent   
5        PK-006   40  199695         HR  2005-02-06               Excellent   

     Days_Absent  
171         10.0  
89           5.0  
194          NaN  
135          7.0  
152         29.0  
..           ...  
131          NaN  
147         22.0  
146          5.0  
17           NaN  
5            NaN  

[200 rows x 7 columns]

Sort the dataset first by 'Department' and then by 'Age' in descending order within each department.

In [13]:
df8 = data.sort_values(by=['Department', 'Age'], ascending=[True, False])
df8

Employee_ID  Age  Salary Department   Join_Date Last_Performance_Rating  \
84       PK-085   64   68360      Admin  2006-08-13                    Good   
95       PK-096   63   53328      Admin  2006-10-29                    Good   
151      PK-152   62  102991      Admin  2007-11-25                    Good   
61       PK-062   61  192688      Admin  2006-03-05               Excellent   
186      PK-187   61   85609      Admin  2008-07-27                    Good   
..          ...  ...     ...        ...         ...                     ...   
162      PK-163   25  135510  Marketing  2008-02-10                 Average   
62       PK-063   23  133727  Marketing  2006-03-12                    Good   
15       PK-016   22  173760  Marketing  2005-04-17                    Good   
47       PK-048   21   69384  Marketing  2005-11-27                     NaN   
94       PK-095   20   68467  Marketing  2006-10-22                 Average   

     Days_Absent  
84           5.0  
95          10.0  
151          4.0  
61           NaN  
186          3.0  
..           ...  
162         11.0  
62          29.0  
15           NaN  
47           NaN  
94          14.0  

[200 rows x 7 columns]

Group the data by 'Department' and find the average 'Salary' for each department.

In [14]:
df9 = data.groupby('Department')['Salary'].mean()
df9

Department
Admin        117073.348837
Finance       98432.592593
HR           122400.028571
IT           122167.894737
Logistics    128903.086957
Marketing    106427.882353
Name: Salary, dtype: float64

Group the data by 'Department' and 'Last_Performance_Rating', then calculate the average 'Age' and the count of employees in each group.

In [15]:
df_10 = data.groupby(['Department', 'Last_Performance_Rating'])['Age'].agg(['mean', 'count'])
df_10

mean  count
Department Last_Performance_Rating                  
Admin      Average                  39.692308     13
           Excellent                49.571429      7
           Good                     47.400000     10
           Outstanding              47.500000      4
           Poor                     33.000000      5
Finance    Average                  52.000000      5
           Excellent                47.500000      2
           Good                     43.454545     11
           Outstanding              38.750000      4
           Poor                     39.500000      2
HR         Average                  51.166667     12
           Excellent                35.833333      6
           Good                     42.000000      9
           Outstanding              44.250000      4
           Poor                     49.000000      2
IT         Average                  33.875000      8
           Excellent                54.750000      8
           Good                     43.000000      9
           Outstanding              32.000000      7
           Poor                     49.750000      4
Logistics  Average                  34.400000      5
           Excellent                38.000000      8
           Good                     43.600000      5
           Outstanding              39.500000      2
           Poor                     46.000000      1
Marketing  Average                  42.285714      7
           Excellent                42.000000      5
           Good                     41.454545     11
           Outstanding              50.000000      2
           Poor                     44.250000      4


### 3. Handling Missing Data:

 Fill all missing values in the 'Days_Absent' column with the average days absent across all employees.

In [16]:
df_11 = data['Days_Absent'].fillna(data['Days_Absent'].mean())
df_11.head()

0     5.00000
1     2.00000
2    12.48125
3    12.48125
4    24.00000
Name: Days_Absent, dtype: float64

Fill missing values in the 'Last_Performance_Rating' with the mode (most frequently occurring value) of that column.

In [17]:
df_12= data['Last_Performance_Rating'].fillna(data['Last_Performance_Rating'].mode()[0])
df_12

0        Excellent
1             Good
2             Poor
3        Excellent
4             Good
          ...     
195        Average
196           Poor
197        Average
198    Outstanding
199           Poor
Name: Last_Performance_Rating, Length: 200, dtype: object

Drop all rows where any data is missing.

In [18]:
df_13 = data.dropna()
df_13

Employee_ID  Age  Salary Department   Join_Date Last_Performance_Rating  \
0        PK-001   58   56736         IT  2005-01-02               Excellent   
1        PK-002   48  142859  Marketing  2005-01-09                    Good   
4        PK-005   27  120084    Finance  2005-01-30                    Good   
6        PK-007   58   37392    Finance  2005-02-13                 Average   
8        PK-009   42   80859         IT  2005-02-27                    Poor   
..          ...  ...     ...        ...         ...                     ...   
195      PK-196   58  144692         HR  2008-09-28                 Average   
196      PK-197   21  117235      Admin  2008-10-05                    Poor   
197      PK-198   22  193165  Logistics  2008-10-12                 Average   
198      PK-199   56  103523  Logistics  2008-10-19             Outstanding   
199      PK-200   36  124476      Admin  2008-10-26                    Poor   

     Days_Absent  
0            5.0  
1            2.0  
4           24.0  
6            4.0  
8            9.0  
..           ...  
195          7.0  
196         13.0  
197         18.0  
198         28.0  
199          5.0  

[145 rows x 7 columns]

Drop any columns where more than 50% of the data is missing (if applicable).

In [19]:
df_14 = data.dropna(axis=1, thresh=len(data)*0.5)
df_14

Employee_ID  Age  Salary Department   Join_Date Last_Performance_Rating  \
0        PK-001   58   56736         IT  2005-01-02               Excellent   
1        PK-002   48  142859  Marketing  2005-01-09                    Good   
2        PK-003   34  142181         IT  2005-01-16                    Poor   
3        PK-004   62  161926  Logistics  2005-01-23               Excellent   
4        PK-005   27  120084    Finance  2005-01-30                    Good   
..          ...  ...     ...        ...         ...                     ...   
195      PK-196   58  144692         HR  2008-09-28                 Average   
196      PK-197   21  117235      Admin  2008-10-05                    Poor   
197      PK-198   22  193165  Logistics  2008-10-12                 Average   
198      PK-199   56  103523  Logistics  2008-10-19             Outstanding   
199      PK-200   36  124476      Admin  2008-10-26                    Poor   

     Days_Absent  
0            5.0  
1            2.0  
2            NaN  
3            NaN  
4           24.0  
..           ...  
195          7.0  
196         13.0  
197         18.0  
198         28.0  
199          5.0  

[200 rows x 7 columns]

<font color ='red'> ======================== Tasks ==========================

Task 0:

Load the elections.csv using Pandas and print its head/tail

In [21]:

df = pd.read_csv('./elections.csv')
df.head()

Year          Candidate                  Party  Popular vote Result  \
0  1824     Andrew Jackson  Democratic-Republican        151271   loss   
1  1824  John Quincy Adams  Democratic-Republican        113142    win   
2  1828     Andrew Jackson             Democratic        642806    win   
3  1828  John Quincy Adams    National Republican        500897   loss   
4  1832     Andrew Jackson             Democratic        702735    win   

           %  
0  57.210122  
1  42.789878  
2  56.203927  
3  43.796073  
4  54.574789

In [29]:
df.describe()

Year  Popular vote           %
count   182.000000  1.820000e+02  182.000000
mean   1934.087912  1.235364e+07   27.470350
std      57.048908  1.907715e+07   22.968034
min    1824.000000  1.007150e+05    0.098088
25%    1889.000000  3.876395e+05    1.219996
50%    1936.000000  1.709375e+06   37.677893
75%    1988.000000  1.897775e+07   48.354977
max    2020.000000  8.126892e+07   61.344703

Task 1:

Write a Pandas command to find all election years where a candidate from the "Democratic" party received more than 50% of the popular vote.

In [49]:
df1 = df.query('Party == "Democratic" and `%` > 50')
df1

Year           Candidate       Party  Popular vote Result          %
2    1828      Andrew Jackson  Democratic        642806    win  56.203927
4    1832      Andrew Jackson  Democratic        702735    win  54.574789
8    1836    Martin Van Buren  Democratic        763291    win  52.272472
13   1844          James Polk  Democratic       1339570    win  50.749477
17   1852     Franklin Pierce  Democratic       1605943    win  51.013168
34   1876    Samuel J. Tilden  Democratic       4288546   loss  51.528376
86   1932  Franklin Roosevelt  Democratic      22821277    win  57.672125
91   1936  Franklin Roosevelt  Democratic      27752648    win  60.978107
94   1940  Franklin Roosevelt  Democratic      27313945    win  54.871202
97   1944  Franklin Roosevelt  Democratic      25612916    win  53.773801
111  1960        John Kennedy  Democratic      34220984    win  50.082561
114  1964      Lyndon Johnson  Democratic      43127041    win  61.344703
123  1976        Jimmy Carter  Democratic      40831881    win  50.271900
162  2008        Barack Obama  Democratic      69498516    win  53.023510
168  2012        Barack Obama  Democratic      65915795    win  51.258484
178  2020        Joseph Biden  Democratic      81268924    win  51.311515

Task 2:

Retrieve the name of the candidate and the result for the election held in 1828 using .iloc.

In [50]:
df2 = df.iloc[3:4,1:5:3]
df2

Candidate Result
3  John Quincy Adams   loss

Task 3:

Use the .query() method to select candidates who lost the election but received over 40% of the vote.

In [51]:
df3 = df.query('Result == "loss" and `%` >= 40')
df3


Year               Candidate                  Party  Popular vote Result  \
0    1824          Andrew Jackson  Democratic-Republican        151271   loss   
3    1828       John Quincy Adams    National Republican        500897   loss   
10   1840        Martin Van Buren             Democratic       1128854   loss   
12   1844              Henry Clay                   Whig       1300004   loss   
14   1848              Lewis Cass             Democratic       1223460   loss   
19   1852          Winfield Scott                   Whig       1386942   loss   
28   1864     George B. McClellan             Democratic       1812807   loss   
29   1868         Horatio Seymour             Democratic       2708744   loss   
31   1872          Horace Greeley     Liberal Republican       2834761   loss   
34   1876        Samuel J. Tilden             Democratic       4288546   loss   
37   1880  Winfield Scott Hancock             Democratic       4444976   loss   
40   1884         James G. Blaine             Republican       4856905   loss   
45   1888        Grover Cleveland             Democratic       5534488   loss   
46   1892       Benjamin Harrison             Republican       5176108   loss   
52   1896  William Jennings Bryan             Democratic       6509052   loss   
55   1900  William Jennings Bryan             Democratic       6370932   loss   
64   1908  William Jennings Bryan             Democratic       6408979   loss   
72   1916    Charles Evans Hughes             Republican       8548728   loss   
83   1928                Al Smith             Democratic      15015464   loss   
96   1940         Wendell Willkie             Republican      22347744   loss   
98   1944         Thomas E. Dewey             Republican      22017929   loss   
104  1948         Thomas E. Dewey             Republican      21991292   loss   
105  1952         Adlai Stevenson             Democratic      27375090   loss   
108  1956         Adlai Stevenson             Democratic      26028028   loss   
112  1960           Richard Nixon             Republican      34108157   loss   
116  1968         Hubert Humphrey             Democratic      31271839   loss   
122  1976             Gerald Ford             Republican      39148634   loss   
129  1980            Jimmy Carter             Democratic      35480115   loss   
134  1984          Walter Mondale             Democratic      37577352   loss   
137  1988         Michael Dukakis             Democratic      41809074   loss   
145  1996                Bob Dole             Republican      39197469   loss   
151  2000                 Al Gore             Democratic      50999897   loss   
158  2004              John Kerry             Democratic      59028444   loss   
166  2008             John McCain             Republican      59948323   loss   
171  2012             Mitt Romney             Republican      60933504   loss   
176  2016         Hillary Clinton             Democratic      65853514   loss   
179  2020            Donald Trump             Republican      74216154   loss   

             %  
0    57.210122  
3    43.796073  
10   46.948787  
12   49.250523  
14   42.552229  
19   44.056548  
28   45.048488  
29   47.334695  
31   44.071406  
34   51.528376  
37   48.278422  
40   48.312208  
45   48.656799  
46   42.984101  
52   46.871053  
55   46.130540  
64   43.414640  
72   46.240779  
83   40.902853  
96   44.894561  
98   46.226199  
104  45.112958  
105  44.446312  
108  42.174464  
112  49.917439  
116  42.863537  
122  48.199499  
129  41.132848  
134  40.729429  
137  45.770691  
145  40.766036  
151  48.491813  
158  48.306775  
166  45.737243  
171  47.384076  
176  48.521539  
179  46.858542

Task 4:

Write a command to sort the dataset first by the year in descending order and then by the percentage of votes in ascending order.

In [52]:
df4 = df.sort_values(by=['Year','%'],ascending=[False,True])
df4


Year          Candidate                  Party  Popular vote Result  \
181  2020     Howard Hawkins                  Green        405035   loss   
180  2020       Jo Jorgensen            Libertarian       1865724   loss   
179  2020       Donald Trump             Republican      74216154   loss   
178  2020       Joseph Biden             Democratic      81268924    win   
172  2016     Darrell Castle           Constitution        203091   loss   
..    ...                ...                    ...           ...    ...   
4    1832     Andrew Jackson             Democratic        702735    win   
3    1828  John Quincy Adams    National Republican        500897   loss   
2    1828     Andrew Jackson             Democratic        642806    win   
1    1824  John Quincy Adams  Democratic-Republican        113142    win   
0    1824     Andrew Jackson  Democratic-Republican        151271   loss   

             %  
181   0.255731  
180   1.177979  
179  46.858542  
178  51.311515  
172   0.149640  
..         ...  
4    54.574789  
3    43.796073  
2    56.203927  
1    42.789878  
0    57.210122  

[182 rows x 6 columns]

Task 5:

Group the data by 'Party' and calculate the average 'Popular vote' and the maximum percentage of votes received by any candidate in that party.

In [57]:
df5 = df.groupby('Party')['Popular vote'].agg(['mean','max'])
df5

mean       max
Party                                        
American               5.156620e+05    873053
American Independent   3.724087e+06   9901118
Anti-Masonic           1.007150e+05    100715
Anti-Monopoly          1.342940e+05    134294
Citizens               2.330520e+05    233052
Communist              1.033070e+05    103307
Constitution           1.821570e+05    203091
Constitutional Union   5.909010e+05    590901
Democratic             2.277206e+07  81268924
Democratic-Republican  1.322065e+05    151271
Dixiecrat              1.175930e+06   1175930
Farmer–Labor           2.653980e+05    265398
Free Soil              2.233555e+05    291501
Green                  8.831137e+05   2882955
Greenback              3.086490e+05    308649
Independent            4.690098e+06  19743821
Liberal Republican     2.834761e+06   2834761
Libertarian            9.554778e+05   4489235
National Democratic    1.346450e+05    134645
National Republican    4.925510e+05    500897
National Union         2.211317e+06   2211317
Natural Law            1.136700e+05    113670
New Alliance           2.172210e+05    217221
Northern Democratic    1.380202e+06   1380202
Populist               4.204167e+05   1041028
Progressive            2.563125e+06   4831706
Prohibition            2.041362e+05    270879
Reform                 4.267094e+06   8085294
Republican             2.616385e+07  74216154
Socialist              4.825871e+05    913693
Southern Democratic    8.480190e+05    848019
States' Rights         1.079290e+05    107929
Taxpayers              1.846560e+05    184656
Union                  8.923780e+05    892378
Union Labor            1.466020e+05    146602
Whig                   1.003282e+06   1386942

Task 6:

Assume there is missing data in the '%' column for some rows. Write a command to fill these missing values with the average percentage of votes received by all candidates in the dataset.

In [60]:
df6 = df['%'].fillna(df['%'].mean())
df6

0      57.210122
1      42.789878
2      56.203927
3      43.796073
4      54.574789
         ...    
177     1.073699
178    51.311515
179    46.858542
180     1.177979
181     0.255731
Name: %, Length: 182, dtype: float64

<font color ='red'> =========== Congrats, Data Scientist keep it up! ==========
<font color ='blue'> ============== The End =============